### ISSUE TITLE
Proposal for Simpler Date/Time Handling

### ISSUE THREAD

**Issue by mark-meyer** [codeforanchorage/courtbot#68]
I tried posting a PR with the updated Twilio yesterday and if failed travis tests because I was setting a date wrong in one of the tests. This made me take a close look at the way we are handling dates and I think we are making everything much harder than it needs to be.

**Here's what we are currently doing**
We are setting Postgres's time zone in `db_connections` to `UTC`. This means that postgres will interpret a date like Jul 22, 2017 2:00pm as `2017-07-22 14:00:00+00`  aka 2pm UTC. 

    SET TIME ZONE 'UTC'  
    SELECT 'July 22, 2017 2:00pm'::timestamptz;
           timestamptz
    ------------------------
     2017-07-22 14:00:00+00

The consequence is that we need to convert *all* of our dates to UTC before inserting them, and do all our date calculations either before or after conversions which is complicated by itself, but more so because of things like daylight savings, leap years, etc.

**Here's what we could do instead**
Set the timezone in Postgres to 'America/Anchorage' in `db_connections`. After that change many things get easier. Postgres will still store dates in UTC but it will read and deliver dates in the correct timezone including daylight savings:

    SET TIME ZONE 'America/Anchorage';
    SELECT 'July 22, 2017 2:00pm'::timestamptz; 
          timestamptz
    ------------------------
     2017-07-22 14:00:00-08

That's the correct time including the the correct utc offset for 2:00pm local time. Dates after DST work as well (note the different offset):

    SELECT 'Nov 22, 2017 2:00pm'::timestamptz;
          timestamptz
    ------------------------
     2017-11-22 14:00:00-09

This means the when we parse the cases from the csv file we no longer need to worry about date conversions. We can simply concat the date and time, insert it, and Postgres will take care of the conversion.

    // ./utils/loaddata.js
    const newCase = {
      date: `${hearingDate} ${c[5]}`
      // etc
      }

This change also makes Postgres's functions like `now()` and  keywords like `today`work as expected so we can lean on them rather than calculating ranges from converted UTC. For example when looking for tomorrow's reminders we can do this (the SQL is read as 'where range contains cases.date'):

    function findReminders() {
    return knex('reminders')
        .where('sent', false)
        .join('cases', 'reminders.case_id', '=', 'cases.id')
        .whereRaw(`tstzrange(TIMESTAMP 'today' + '1 day', TIMESTAMP 'today' + '2 days') @> ${cases.date}`)
        .select();
    }

instead of:

    function findReminders() {
    // database is converting dates to UTC, so we need to
    // convert our comparison date to UTC prior to comparing.
    const todayMidnight = dates.now().add('1', 'days').hour(0).minute(0)
        .utcOffset(0)
        .format();
    const tomorrowMidnight = dates.now().add('2', 'days').hour(0).minute(0)
        .utcOffset(0)
        .format();
    return knex('reminders')
        .where('sent', false)
        .join('cases', 'reminders.case_id', '=', 'cases.id')
        .whereRaw(`(date (cases.date at time zone '${process.env.TIMEZONE}') <= date '${tomorrowMidnight}' at time zone '${process.env.TIMEZONE}')
                    AND (date (cases.date at time zone '${process.env.TIMEZONE}') > date '${todayMidnight}' at time zone '${process.env.TIMEZONE}')`)
        .select();
    } 

There are some places where we want to know if a date is today or has already happened, such as when responding to a reminder request. Here also we can also let Postgres do the work by creating the column from a function. So `findCitation()` could look like this.

    function findCitation(citation) {
    const cleanedCitation = escapeSQL(citation.toUpperCase().trim());
    return knex('cases').where('citations', '@>', `[{"id": "${cleanedCitation}"}]`)
    .select('*', knex.raw(
        `tstzrange(TIMESTAMP 'today', TIMESTAMP 'today' + interval '1 day') @> date as today,
        date < now() as has_past
        `));
    }

The returned rows will have boolean columns for `today` and `has_past` which can be used when we create the message like:

    if (match.has_past) {
        futureHearing = ' a future hearing';

When we add reminders we can use Postgres's `now()` for `created_at` instead of calculating the current UTC:

    created_at: knex.raw('now()'),

and it will do the right thing. 

If we do this, we will be out of the time conversion business altogether — `./utils/dates/` can go away. We can still use `moment` for formatting without any trouble but we wouldn't need to ever give it a timezone or offset it would purely be formatting local time. 

**The primary difficulty with this**  
Getting the app working wasn't too hard, but we will need to change *a lot* of tests. Because we can't (I don't think) freeze time on the DB we can't do things like insert a case in 2015 and ask if it's today -- that calculation will be made by Postgres not our date-frozen code. Those kind of tests will require setting fixtures to dates relative to the test. The upside is that the test might be easier read.

**Does anyone think this is a bad idea?**

**Comment by mfrederickson** [comment]
Sounds great.  Very thorough analysis.

**Comment by tescher** [comment]
Sounds better, but again it isn't something I would want done in the main
stream while we are trying to go live.

Tim Escher
tim@eschers.com
608 432 5799

On Sep 23, 2017 1:53 PM, "Marvin Frederickson" <notifications@github.com>
wrote:

Sounds great. Very thorough analysis.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/codeforanchorage/courtbot/issues/68#issuecomment-331662585>,
or mute the thread

**Comment by markmeyerphoto** [comment]
I think that's reasonable Tim - it will be a pretty big change. Maybe can start something akin to a v.2 branch that includes this with revamped criminal case handling and DB changes.

One thing I didn't notice is that `moment().format()` by itself formats dates relative to the server's timezone. So if I set node's TZ to something exotic like `America/St_Johns` this:

    moment('2017-09-26T13:30:00-08:00').format('dddd, MMM Do H:mm')

returns: 'Tuesday, Sep 26th 19:00', which is not what we want.

Setting `process.env.TZ = 'AMERICA/ANCHORAGE' ` fixes that for me, but I'm not sure if that's dependable on all platforms.

**Comment by tescher** [comment]
I like V2 for all that. Also, we might want to explore if there is a way to
integrate this into the legal access portal project Microsoft is starting
up here. Brendan would have more info on that

Tim Escher
tim@eschers.com
608-432-5799

**Comment by brendanbabb** [comment]
The Portal is aimed at civil law currently, but I could see this project possibly being useful for the portal.

**Comment by markmeyerphoto** [comment]
I have pushed a branch called `update_libs` which has these date/time changes. It also updates all the dependencies and gets rid of a few that weren't used. Some of the updates like Express and Twillio had substantial breaking changes so code and tests needed to be adjusted. The only out-of-date library left is Rollbar (current: 0.6.6 - latest: 2.2.8). I imagine I'll incorporating this into a logging module so I'll update then and make the necessary changes requires by its updated API.  

I think if we want to push ahead with V2 changes, this will be a good foundation — we were going to have to update some of this sooner or later. The changes to the Date/Time code would probably benefit from a review by someone in case I've overlooked some cases.
